crates/api/src/values.rs

```diff
@@ -6,21 +6,71 @@ use std::ptr;
 use wasmtime_environ::ir;
 use wasmtime_jit::RuntimeValue;
 
+/// Possible runtime values that a WebAssembly module can either consume or
+/// produce.
```

Comment:
Perhaps we should describe here and in `RuntimeValue` what is the difference between these `Val` and `RuntimeValue`. I'm not very clear on their differences and judging superficially from some of the overlap (e.g. `RuntimeValue::unwrap_*`) I wonder if we shouldn't just create one version of a runtime value struct and move it to a place where both crates can share the definition.

Reply:
Hm I'm not sure if this would be the best place for that, but it sort of depends on the user. The `wasmtime` crate is still relatively new and idioms are still changing, but I believe the intention is that the `wasmtime` crate is almost exclusively used when embedding wasmtime somewhere else, unless a much more advanced use case is required.

The `RuntimeValue` type, AFAIK, is purely wasmtime-internals in that it's part of the nitty-gritty details of `wasmtime`'s own internal implementation. The `Val` type here, though, is intended to be user-facing and what you actually interact with. I don't think users are expected to interact with `RuntimeValue` at all if they're using the `wasmtime` crate. 

So basically I think that this won't be a problem as time goes on because users won't ever even see `RuntimeValue`, and otherwise proposals like https://github.com/bytecodealliance/wasmtime/issues/677 may make unifying impossible since the `wasmtime::Val` will be richer than `RuntimeValue`.